bin/monitor.ml

```diff
@@ -0,0 +1,230 @@
+  ;;
+
+  let render_state (state : t) =
+    Dune_rpc.Diagnostic.Set.iter state.diagnostics ~f:(fun d ->
+      Console.print_user_message (Dune_rpc.Diagnostic.to_user_message d));
+    status state
+  ;;
+
+  let maybe_reset_diagnostics (state : t) =
+    match (state.progress : Dune_rpc.Progress.t) with
+    | Interrupted ->
+      state.diagnostics <- init.diagnostics;
+      Console.reset ();
+      render_state state
+    | _ -> status state
+  ;;
+end
+
+(** A generic loop that fetches events from a [poller] and writes them to the
+    [event] mvar. *)
+let fetch_loop (event : Event.t Fiber.Mvar.t) (poller : 'a Client.Stream.t) ~f =
+  let rec loop () =
+    let* next = Client.Stream.next poller in
+    match next with
+    | Some payload -> Fiber.Mvar.write event (f payload) >>> loop ()
+    | None -> Fiber.Mvar.write event Event.Poller_disconnected
+  in
+  loop ()
+;;
+
+(** The main render loop. Reads the [event] mvar and updates the display. We
+    need to make sure to always clear the events in the mvar or else the other
+    loops will be stuck writing. We also pass the [client] so we can ping it
+    when we lose connection. *)
+let render_loop client event =
+  let state = State.init in
+  let rec loop () =
+    let* event = Fiber.Mvar.read event in
+    match event with
+    (* If we are quitting then just return. *)
+    | _ when state.quit -> Fiber.return ()
+    (* Update internal job counter on job events. *)
+    | Event.Jobs jobs ->
+      List.iter jobs ~f:(fun jobs ->
+        match (jobs : Dune_rpc.Job.Event.t) with
+        | Start _ -> state.job_number <- state.job_number + 1
+        | Stop _ -> state.job_number <- state.job_number - 1);
+      loop ()
+    (* Update the progress status. If the progress is [Interuptted] then reset
+       the diagnostics and rerender the window. *)
+    | Event.Progress progress ->
+      state.progress <- progress;
+      State.maybe_reset_diagnostics state;
+      loop ()
+    (* Update the diagnostics. If diagnostics are added then print them. If
+       they are removed then udpate the state and rerender entirely. *)
+    | Event.Diagnostics diagnostics ->
+      let module Diagnostic = Dune_rpc.Diagnostic in
+      List.iter diagnostics ~f:(function
+        | Diagnostic.Event.Add d ->
+          state.diagnostics <- Diagnostic.Set.add state.diagnostics d;
+          Console.print_user_message (Diagnostic.to_user_message d);
+          State.status state (* we make sure to update the status here *)
+        | Remove d ->
+          state.diagnostics <- Diagnostic.Set.remove state.diagnostics d;
+          Console.reset ();
+          State.render_state state);
+      loop ()
+    (* If the poller disconnected in one of the writer loops then ping the
+       server. If it is not responding then we know the server must have
+       disconnected and we can quit. *)
+    | Poller_disconnected ->
+      (* ping the server to make sure it's still alive *)
```

Comment:
Is this even possible?

Reply:
This happens when the server shuts down for instance. If I have a watch mode dune and a monitor and I shut down the watch mode dune, then the fetch loops will raise an exception about the poller. That then gets turned into a `Poller_disconnected` event which we process here.

Reply:
Okay, but then how will the ping succeed?

Reply:
Ah I see what you mean. How else can I be sure that the server has disconnected and I am not getting some other error from the poll.

Reply:
If there's another error, you can't really recover from it anyway. So there's no difference and you should terminate the connection regardless.

You could add a `--auto-poll` mode that tries to re-establish the connection if it's dead.

Reply:
I'll leave that for a future PR for now.